File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        seen = {}
        left = 0
        best = 0
        for right, ch in enumerate(s):
            if ch in seen and seen[ch] >= left:
                left = seen[ch] + 1
            seen[ch] = right
            cur = right - left + 1
            if cur > best:
                best = cur
        return best

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // last position seen for each ASCII char; -1 = unseen
        vector<int> last(256, -1);
        int best = 0, left = 0;                 // window = [left, i]
        for (int i = 0; i < (int)s.size(); ++i) {
            unsigned char c = s[i];
            if (last[c] >= left) left = last[c] + 1; // move left past duplicate
            last[c] = i;
            best = max(best, i - left + 1);
        }
        return best;
    }
};
